File: core-rust/src/handlers/payments.rs
use actix_web::{web, HttpResponse};
use rust_decimal::Decimal;
use sqlx::PgPool;
use uuid::Uuid;

use crate::models::payment::*;

/// GET /api/payments/
#[utoipa::path(
    get,
    path = "/api/payments/",
    tag = "payments",
    params(
        ("user_id" = Option<Uuid>, Query, description = "Filter by user ID"),
        ("payment_type" = Option<String>, Query, description = "Filter by payment type"),
        ("status" = Option<String>, Query, description = "Filter by status")
    ),
    responses(
        (status = 200, description = "List of payments")
    )
)]
pub async fn list_payments(
    pool: web::Data<PgPool>,
    query: web::Query<PaymentQuery>,
) -> HttpResponse {
    let payments = match (&query.user_id, &query.payment_type, &query.status) {
        (Some(uid), Some(pt), Some(st)) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE user_id=$1 AND payment_type=$2 AND status=$3 ORDER BY created_at DESC",
            )
            .bind(uid)
            .bind(pt)
            .bind(st)
            .fetch_all(pool.get_ref())
            .await
        }
        (Some(uid), Some(pt), None) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE user_id=$1 AND payment_type=$2 ORDER BY created_at DESC",
            )
            .bind(uid)
            .bind(pt)
            .fetch_all(pool.get_ref())
            .await
        }
        (Some(uid), None, Some(st)) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE user_id=$1 AND status=$2 ORDER BY created_at DESC",
            )
            .bind(uid)
            .bind(st)
            .fetch_all(pool.get_ref())
            .await
        }
        (Some(uid), None, None) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE user_id=$1 ORDER BY created_at DESC",
            )
            .bind(uid)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, Some(pt), Some(st)) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE payment_type=$1 AND status=$2 ORDER BY created_at DESC",
            )
            .bind(pt)
            .bind(st)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, Some(pt), None) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE payment_type=$1 ORDER BY created_at DESC",
            )
            .bind(pt)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, None, Some(st)) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments WHERE status=$1 ORDER BY created_at DESC",
            )
            .bind(st)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, None, None) => {
            sqlx::query_as::<_, Payment>(
                "SELECT * FROM payments ORDER BY created_at DESC LIMIT 100",
            )
            .fetch_all(pool.get_ref())
            .await
        }
    };

    match payments {
        Ok(p) => HttpResponse::Ok().json(serde_json::json!({"results": p})),
        Err(e) => {
            tracing::error!("Failed to fetch payments: {}", e);
            HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}))
        }
    }
}

/// GET /api/payments/{id}/
#[utoipa::path(
    get,
    path = "/api/payments/{id}/",
    tag = "payments",
    params(("id" = i32, Path, description = "Payment ID")),
    responses(
        (status = 200, description = "Payment details", body = Payment),
        (status = 404, description = "Payment not found")
    )
)]
pub async fn get_payment(pool: web::Data<PgPool>, path: web::Path<i32>) -> HttpResponse {
    let payment_id = path.into_inner();
    match sqlx::query_as::<_, Payment>("SELECT * FROM payments WHERE id = $1")
        .bind(payment_id)
        .fetch_optional(pool.get_ref())
        .await
    {
        Ok(Some(p)) => HttpResponse::Ok().json(p),
        Ok(None) => HttpResponse::NotFound().json(serde_json::json!({"detail": "Not found."})),
        Err(e) => {
            tracing::error!("Failed to fetch payment: {}", e);
            HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}))
        }
    }
}

/// POST /api/payments/{id}/simulate_success/
#[utoipa::path(
    post,
    path = "/api/payments/{id}/simulate_success/",
    tag = "payments",
    params(("id" = i32, Path, description = "Payment ID")),
    responses(
        (status = 200, description = "Payment simulated as successful"),
        (status = 404, description = "Payment not found")
    )
)]
pub async fn simulate_success(pool: web::Data<PgPool>, path: web::Path<i32>) -> HttpResponse {
    let payment_id = path.into_inner();

    let payment = match sqlx::query_as::<_, Payment>("SELECT * FROM payments WHERE id = $1")
        .bind(payment_id)
        .fetch_optional(pool.get_ref())
        .await
    {
        Ok(Some(p)) => p,
        Ok(None) => return HttpResponse::NotFound().json(serde_json::json!({"detail": "Not found."})),
        Err(e) => {
            tracing::error!("Failed to fetch payment: {}", e);
            return HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}));
        }
    };

    // Mark payment as succeeded
    let _ = sqlx::query(
        "UPDATE payments SET status='succeeded', paid_at=NOW(), updated_at=NOW() WHERE id=$1",
    )
    .bind(payment_id)
    .execute(pool.get_ref())
    .await;

    // Update user balance
    let _ = sqlx::query(
        "UPDATE users SET balance = balance + $1, updated_at=NOW() WHERE id = $2",
    )
    .bind(payment.amount)
    .bind(payment.user_id)
    .execute(pool.get_ref())
    .await;

    // Record transaction
    let new_balance: Decimal = sqlx::query_scalar("SELECT balance FROM users WHERE id = $1")
        .bind(payment.user_id)
        .fetch_one(pool.get_ref())
        .await
        .unwrap_or(Decimal::ZERO);

    let _ = sqlx::query(
        r#"INSERT INTO transactions (user_id, transaction_type, amount, balance_after, payment_id, description)
           VALUES ($1, 'deposit', $2, $3, $4, 'Payment simulation')"#,
    )
    .bind(payment.user_id)
    .bind(payment.amount)
    .bind(new_balance)
    .bind(payment_id)
    .execute(pool.get_ref())
    .await;

    HttpResponse::Ok().json(serde_json::json!({"status": "succeeded", "payment_id": payment_id}))
}

/// GET /api/payments/transactions/
#[utoipa::path(
    get,
    path = "/api/payments/transactions/",
    tag = "payments",
    params(
        ("user_id" = Option<Uuid>, Query, description = "Filter by user ID"),
        ("transaction_type" = Option<String>, Query, description = "Filter by transaction type")
    ),
    responses(
        (status = 200, description = "List of transactions")
    )
)]
pub async fn list_transactions(
    pool: web::Data<PgPool>,
    query: web::Query<TransactionQuery>,
) -> HttpResponse {
    let transactions = match (&query.user_id, &query.transaction_type) {
        (Some(uid), Some(tt)) => {
            sqlx::query_as::<_, Transaction>(
                "SELECT * FROM transactions WHERE user_id=$1 AND transaction_type=$2 ORDER BY created_at DESC",
            )
            .bind(uid)
            .bind(tt)
            .fetch_all(pool.get_ref())
            .await
        }
        (Some(uid), None) => {
            sqlx::query_as::<_, Transaction>(
                "SELECT * FROM transactions WHERE user_id=$1 ORDER BY created_at DESC",
            )
            .bind(uid)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, Some(tt)) => {
            sqlx::query_as::<_, Transaction>(
                "SELECT * FROM transactions WHERE transaction_type=$1 ORDER BY created_at DESC",
            )
            .bind(tt)
            .fetch_all(pool.get_ref())
            .await
        }
        (None, None) => {
            sqlx::query_as::<_, Transaction>(
                "SELECT * FROM transactions ORDER BY created_at DESC LIMIT 100",
            )
            .fetch_all(pool.get_ref())
            .await
        }
    };

    match transactions {
        Ok(t) => HttpResponse::Ok().json(serde_json::json!({"results": t})),
        Err(e) => {
            tracing::error!("Failed to fetch transactions: {}", e);
            HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}))
        }
    }
}

/// GET /api/payments/transactions/{id}/
#[utoipa::path(
    get,
    path = "/api/payments/transactions/{id}/",
    tag = "payments",
    params(("id" = i32, Path, description = "Transaction ID")),
    responses(
        (status = 200, description = "Transaction details", body = Transaction),
        (status = 404, description = "Transaction not found")
    )
)]
pub async fn get_transaction(pool: web::Data<PgPool>, path: web::Path<i32>) -> HttpResponse {
    let tx_id = path.into_inner();
    match sqlx::query_as::<_, Transaction>("SELECT * FROM transactions WHERE id = $1")
        .bind(tx_id)
        .fetch_optional(pool.get_ref())
        .await
    {
        Ok(Some(t)) => HttpResponse::Ok().json(t),
        Ok(None) => HttpResponse::NotFound().json(serde_json::json!({"detail": "Not found."})),
        Err(e) => {
            tracing::error!("Failed to fetch transaction: {}", e);
            HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}))
        }
    }
}

/// GET /api/payments/transactions/summary/
#[utoipa::path(
    get,
    path = "/api/payments/transactions/summary/",
    tag = "payments",
    params(("user_id" = Uuid, Query, description = "User ID")),
    responses(
        (status = 200, description = "Transaction summary", body = TransactionSummary),
        (status = 400, description = "user_id is required")
    )
)]
pub async fn transaction_summary(
    pool: web::Data<PgPool>,
    query: web::Query<std::collections::HashMap<String, String>>,
) -> HttpResponse {
    let user_id = match query.get("user_id").and_then(|s| Uuid::parse_str(s).ok()) {
        Some(id) => id,
        None => return HttpResponse::BadRequest().json(serde_json::json!({"error": "user_id is required"})),
    };

    let current_balance: Decimal = sqlx::query_scalar("SELECT balance FROM users WHERE id = $1")
        .bind(user_id)
        .fetch_optional(pool.get_ref())
        .await
        .unwrap_or(None)
        .unwrap_or(Decimal::ZERO);

    let total_deposited: Decimal = sqlx::query_scalar(
        "SELECT COALESCE(SUM(amount), 0) FROM transactions WHERE user_id=$1 AND transaction_type='deposit'",
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await
    .unwrap_or(Decimal::ZERO);

    let total_withdrawn: Decimal = sqlx::query_scalar(
        "SELECT COALESCE(SUM(amount), 0) FROM transactions WHERE user_id=$1 AND transaction_type='withdrawal'",
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await
    .unwrap_or(Decimal::ZERO);

    let total_refunded: Decimal = sqlx::query_scalar(
        "SELECT COALESCE(SUM(amount), 0) FROM transactions WHERE user_id=$1 AND transaction_type='refund'",
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await
    .unwrap_or(Decimal::ZERO);

    let transaction_count: i64 = sqlx::query_scalar(
        "SELECT COUNT(*) FROM transactions WHERE user_id=$1",
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await
    .unwrap_or(0);

    HttpResponse::Ok().json(TransactionSummary {
        current_balance,
        total_deposited,
        total_withdrawn,
        total_refunded,
        transaction_count,
    })
}

/// POST /api/payments/webhook/
pub async fn payment_webhook(body: web::Bytes) -> HttpResponse {
    // Auto-detect provider from payload structure and route accordingly
    let payload = match serde_json::from_slice::<serde_json::Value>(&body) {
        Ok(v) => v,
        Err(_) => return HttpResponse::BadRequest().json(serde_json::json!({"error": "Invalid JSON"})),
    };

    // Detect provider: Tochka uses "event" field with "payment." prefix; YooKassa uses "type" field
    let is_tochka = payload.get("event").is_some();
    let provider = if is_tochka { "tochka" } else { "yookassa" };

    tracing::info!("Received payment webhook from provider: {}", provider);

    // In production this would verify webhook signatures and update payment status
    HttpResponse::Ok().json(serde_json::json!({"received": true, "provider": provider}))
}

/// POST /api/payments/
#[utoipa::path(
    post,
    path = "/api/payments/",
    tag = "payments",
    request_body = CreatePayment,
    responses(
        (status = 201, description = "Payment created", body = Payment),
        (status = 400, description = "Bad request")
    )
)]
pub async fn create_payment(
    pool: web::Data<PgPool>,
    body: web::Json<CreatePayment>,
) -> HttpResponse {
    let data = body.into_inner();
    let description = data.description.unwrap_or_default();

    match sqlx::query_as::<_, Payment>(
        r#"INSERT INTO payments (user_id, payment_type, amount, procurement_id, description)
           VALUES ($1, $2, $3, $4, $5)
           RETURNING *"#,
    )
    .bind(data.user_id)
    .bind(&data.payment_type)
    .bind(data.amount)
    .bind(data.procurement_id)
    .bind(&description)
    .fetch_one(pool.get_ref())
    .await
    {
        Ok(payment) => HttpResponse::Created().json(payment),
        Err(e) => {
            tracing::error!("Failed to create payment: {}", e);
            HttpResponse::BadRequest().json(serde_json::json!({"error": format!("{}", e)}))
        }
    }
}

/// GET /api/payments/{id}/status/
#[utoipa::path(
    get,
    path = "/api/payments/{id}/status/",
    tag = "payments",
    params(("id" = i32, Path, description = "Payment ID")),
    responses(
        (status = 200, description = "Payment status", body = PaymentStatusResponse),
        (status = 404, description = "Payment not found")
    )
)]
pub async fn get_payment_status(pool: web::Data<PgPool>, path: web::Path<i32>) -> HttpResponse {
    let payment_id = path.into_inner();
    match sqlx::query_as::<_, Payment>("SELECT * FROM payments WHERE id = $1")
        .bind(payment_id)
        .fetch_optional(pool.get_ref())
        .await
    {
        Ok(Some(payment)) => {
            let status_display = match payment.status.as_str() {
                "pending" => "Pending",
                "waiting_for_capture" => "Waiting for Capture",
                "succeeded" => "Succeeded",
                "cancelled" => "Cancelled",
                "refunded" => "Refunded",
                other => other,
            }
            .to_string();

            HttpResponse::Ok().json(PaymentStatusResponse {
                id: payment.id,
                status: payment.status,
                status_display,
                amount: payment.amount,
                confirmation_url: payment.confirmation_url,
            })
        }
        Ok(None) => HttpResponse::NotFound().json(serde_json::json!({"detail": "Not found."})),
        Err(e) => {
            tracing::error!("Failed to fetch payment: {}", e);
            HttpResponse::InternalServerError().json(serde_json::json!({"error": "Database error"}))
        }
    }
}

File: core-fastapi/app/routers/payments.py
"""Payment endpoints — mirrors core-rust handlers/payments.rs"""

import logging
from decimal import Decimal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query

from ..db import get_pool
from ..schemas import CreatePayment, CreateWithdrawal

logger = logging.getLogger("core.payments")
router = APIRouter(prefix="/payments", tags=["payments"])


@router.post("/", status_code=201, summary="Create payment")
async def create_payment(body: CreatePayment, pool=Depends(get_pool)):
    try:
        row = await pool.fetchrow(
            """INSERT INTO payments (user_id, payment_type, amount, procurement_id, description)
               VALUES ($1,$2,$3,$4,$5)
               RETURNING *""",
            body.user_id,
            body.payment_type,
            body.amount,
            body.procurement_id,
            body.description or "",
        )
        return dict(row)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/", summary="List payments for a user")
async def list_payments(
    user_id: UUID | None = Query(default=None),
    pool=Depends(get_pool),
):
    if user_id is not None:
        rows = await pool.fetch(
            "SELECT * FROM payments WHERE user_id=$1 ORDER BY created_at DESC LIMIT 200",
            user_id,
        )
    else:
        rows = await pool.fetch(
            "SELECT * FROM payments ORDER BY created_at DESC LIMIT 200"
        )
    return {"results": [dict(r) for r in rows]}


@router.get("/{payment_id}/status/", summary="Get payment status")
async def get_payment_status(payment_id: int, pool=Depends(get_pool)):
    row = await pool.fetchrow("SELECT * FROM payments WHERE id=$1", payment_id)
    if not row:
        raise HTTPException(status_code=404, detail="Not found.")
    status_display = {
        "pending": "Pending",
        "waiting_for_capture": "Waiting for Capture",
        "succeeded": "Succeeded",
        "cancelled": "Cancelled",
        "refunded": "Refunded",
    }.get(row["status"], row["status"])
    return {
        "id": row["id"],
        "status": row["status"],
        "status_display": status_display,
        "amount": row["amount"],
        "confirmation_url": row["confirmation_url"],
    }


# ─── Withdrawal requests (issue #194: form 1.4 «Вывод средств») ───────────────

@router.post("/withdrawals/", status_code=201, summary="Request a withdrawal")
async def create_withdrawal(body: CreateWithdrawal, pool=Depends(get_pool)):
    if body.amount <= 0:
        raise HTTPException(status_code=400, detail="amount must be positive.")
    if not body.bank_details.strip():
        raise HTTPException(
            status_code=400,
            detail={"bank_details": ["Обязательное поле."]},
        )
    user = await pool.fetchrow("SELECT id, balance FROM users WHERE id=$1", body.user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found.")
    if user["balance"] < body.amount:
        raise HTTPException(status_code=400, detail="Insufficient funds.")
    row = await pool.fetchrow(
        """INSERT INTO withdrawal_requests (user_id, amount, bank_details, status)
           VALUES ($1, $2, $3, 'pending') RETURNING *""",
        body.user_id, body.amount, body.bank_details.strip(),
    )
    return dict(row)


@router.get("/withdrawals/", summary="List withdrawal requests")
async def list_withdrawals(
    user_id: UUID | None = Query(default=None),
    status: str | None = Query(default=None),
    pool=Depends(get_pool),
):
    clauses: list[str] = []
    args: list = []
    if user_id is not None:
        args.append(user_id)
        clauses.append(f"user_id=${len(args)}")
    if status:
        args.append(status)
        clauses.append(f"status=${len(args)}")
    where = ("WHERE " + " AND ".join(clauses)) if clauses else ""
    rows = await pool.fetch(
        f"SELECT * FROM withdrawal_requests {where} ORDER BY created_at DESC LIMIT 200",
        *args,
    )
    return {"results": [dict(r) for r in rows]}


@router.post("/withdrawals/{withdrawal_id}/process/", summary="Process a withdrawal request")
async def process_withdrawal(withdrawal_id: int, body: dict | None = None, pool=Depends(get_pool)):
    """Marks a withdrawal as approved or rejected. On approval the amount is
    debited from the user's balance and recorded in ``transactions``."""
    new_status = (body or {}).get("status", "approved")
    if new_status not in ("approved", "rejected"):
        raise HTTPException(status_code=400, detail="status must be 'approved' or 'rejected'.")

    async with pool.acquire() as conn:
        async with conn.transaction():
            wr = await conn.fetchrow(
                "SELECT * FROM withdrawal_requests WHERE id=$1 FOR UPDATE", withdrawal_id
            )
            if not wr:
                raise HTTPException(status_code=404, detail="Not found.")
            if wr["status"] != "pending":
                raise HTTPException(status_code=400, detail="Already processed.")

            if new_status == "approved":
                user = await conn.fetchrow(
                    "SELECT balance FROM users WHERE id=$1 FOR UPDATE", wr["user_id"]
                )
                if not user or user["balance"] < wr["amount"]:
                    raise HTTPException(status_code=400, detail="Insufficient funds.")
                new_balance = user["balance"] - wr["amount"]
                await conn.execute(
                    "UPDATE users SET balance=$2, updated_at=NOW() WHERE id=$1",
                    wr["user_id"], new_balance,
                )
                await conn.execute(
                    """INSERT INTO transactions
                         (user_id, transaction_type, amount, balance_after, description)
                       VALUES ($1, 'withdrawal', $2, $3, 'Withdrawal request approved')""",
                    wr["user_id"], -wr["amount"], new_balance,
                )

            row = await conn.fetchrow(
                """UPDATE withdrawal_requests
                      SET status=$2, processed_at=NOW()
                    WHERE id=$1 RETURNING *""",
                withdrawal_id, new_status,
            )
    return dict(row)
